### ISSUE TITLE
Aggregate loss mean makes float64 from integers

### ISSUE THREAD

**Issue by Botev** [Lasagne/Lasagne#472]
The following code:
```
if mode == 'mean':
        loss = loss.mean()
        return loss
    elif mode == 'sum':
        return loss.sum()
```
And I think some code around it can produce float64 if it is called with a loss which is integer type (e.g. accuracy). This can raise error if one has the "warn_float64" option on theano. I was wondering if it would be worth to just change it to `loss.mean(dtype=theano.config.floatX)` to make it always produce the correct precision.

**Comment by f0k** [comment]
Difficult question, because "correct precision" can depend on the use case. What about `loss.sum()`, for example, should it also return a float for an integer loss? And should we return a float32 mean if loss is float64 already? Maybe it would be better to cast the loss to `floatX` outside of `aggregate()` if that's what you want. I'm all up for discussing this, though.

**Comment by f0k** [comment]
> Maybe it would be better to cast the loss to `floatX` outside of `aggregate()` if that's what you want.

Comparing the two variants:
```
togpu = theano.sandbox.cuda.as_cuda_ndarray_variable
a = togpu(T.matrix('a'))
b = T.ivector('b')
l = T.eq(T.argmax(a, axis=1), b)
# floatX mean of integer loss:
fn1 = theano.function([a,b], togpu(l.mean(dtype=theano.config.floatX)))
theano.printing.debugprint(fn1)
GpuElemwise{TrueDiv}[(0, 0)] [@A] ''   8
 |GpuFromHost [@B] ''   7
 | |Sum{acc_dtype=int64} [@C] ''   6
 |   |Elemwise{eq,no_inplace} [@D] ''   5
 |     |MaxAndArgmax.1 [@E] 'argmax'   3
 |     | |HostFromGpu [@F] ''   1
 |     | | |<CudaNdarrayType(float32, matrix)> [@G]
 |     | |TensorConstant{(1,) of 1} [@H]
 |     |b [@I]
 |GpuFromHost [@J] ''   4
   |Elemwise{Cast{float32}} [@K] ''   2
     |Shape_i{0} [@L] ''   0
       |<CudaNdarrayType(float32, matrix)> [@G]

# cast loss to floatX before:
fn2 = theano.function([a,b], togpu(l.astype(theano.config.floatX).mean()))
theano.printing.debugprint(fn2)
GpuElemwise{TrueDiv}[(0, 0)] [@A] ''   8
 |GpuCAReduce{add}{1} [@B] ''   7
 | |GpuFromHost [@C] ''   6
 |   |Elemwise{Composite{Cast{float32}(EQ(i0, i1))}} [@D] ''   5
 |     |MaxAndArgmax.1 [@E] 'argmax'   3
 |     | |HostFromGpu [@F] ''   1
 |     | | |<CudaNdarrayType(float32, matrix)> [@G]
 |     | |TensorConstant{(1,) of 1} [@H]
 |     |b [@I]
 |GpuFromHost [@J] ''   4
   |Elemwise{Cast{float32}} [@K] ''   2
     |Shape_i{0} [@L] ''   0
       |<CudaNdarrayType(float32, matrix)> [@G]
```

So actually, `T.argmax` and `T.eq` is performed on the CPU, even if we pretend you want the end result on the GPU. Summarization happens on CPU for the first variant, and on GPU for the second. In any case, things are copied back to GPU for dividing the sum by the number of data points.

I guess we should leave everything as it is. If you want to compute an accuracy, you probably want to output it, so you need it on CPU and it's totally fine to compute it in float64 precision instead of copying things back to GPU. If you want to compute something you want to calculate a gradient of, you probably want to cast the loss to float32. We can think about adding a `dtype` argument to `mean()` once there is a use case that needs it, but I think we shouldn't change it just in case somebody has the float64 warning enabled.

I'm closing this so we don't forget, but please feel free to comment and/or reopen if you disagree!

**Comment by Botev** [comment]
I think opting for the dtype might worth. If by for some reason your ouput is integer based. However, not sure if there is a usecase for it.
